### PR TITLE
New implementation of RPS load with periodic workload process.

### PR DESCRIPTION
**Pull request by amikryukov**
Invocations starts with calculated period.
If thread pull is full (not enough threads to provide specified rps) - WARN message occurs.

### CODE REVIEW

**Comment by amikryukov** [comment]
PerThreadWorkloadProcess is remain the same (not using AbstractWorkloadProcess) by purpose.

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/PeriodSingleTaskScheduler.java [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+
+
+    Logger log = LoggerFactory.getLogger(PeriodSingleTaskScheduler.class);
+
+
+    public PeriodSingleTaskScheduler() {
+        this(DEFAULT_CORE_POOL_SIZE);
+    }
+
+    public PeriodSingleTaskScheduler(int corePoolSize) {
```
Comment (kgribov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/PeriodSingleTaskScheduler.java]: for what purposes you need corePoolSize ?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/PeriodSingleTaskScheduler.java [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+     * @param initialDelay initialDelay before executing command
+     * @param period period
+     * @param unit time unit for period, initialDelay
+     */
+    public void scheduleAtFixedRate(final Runnable command,
+                                    long initialDelay,
+                                    long period,
+                                    TimeUnit unit) {
+
+        lock.lock();
```
Comment (kgribov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/PeriodSingleTaskScheduler.java]: Why not to use sync methods ?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+                    };
+
+                    Callable<State> foo = new Callable<State>() {
+                        @Override
+                        public State call() throws Exception {
+                            try {
+                                // wait till execution stops
+                                lfs.get();
+                            } catch (InterruptedException e) {
+                                e.printStackTrace();
```
Comment (nmusienko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java]: WAT?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java [hunk 4]
```diff
@@ -0,0 +1,407 @@
...
+                try {
+                    return finishedSamples.get() + delegate.getFinishedSamples();
+                } finally {
+                    lock.unlock();
+                }
+            }
+
+            @Override
+            public boolean isRunning() {
+                lock.lock();
```
Comment (nmusienko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java]: overhead...

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadProcess.java [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+                .addCollectors(collectors)
+                .addValidators(validators)
+                .addListeners(listeners)
+                .useExecutor(executor);
+        WorkloadService thread = getService(builder);
+        thread.changeDelay(delay);
+        log.debug("Starting workload");
+        Future<Service.State> future = thread.start();
+        Service.State state = Futures.get(future, timeoutsConfiguration.getWorkloadStartTimeout());
+        log.debug("Workload thread with is started with state {}", state);
```
Comment (nmusienko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadProcess.java]: typo

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java [hunk 6]
```diff
@@ -0,0 +1,407 @@
...
+            ImmutableList<Validator> validatorList = validators.build();
+            ImmutableList<InvocationListener> listenersList = listeners.build();
+
+            scenario.setInvocationListener(Invokers.validateListener(validatorList, (Iterable) collectorsList, (List)listenersList));
+            return new PredefinedSamplesWorkloadService(Invokers.mergeFlushElements(validatorList, collectorsList), samples);
+        }
+
+        public WorkloadService buildServiceWithSharedSamplesCount(final AtomicInteger samples) {
+            ImmutableList<ScenarioCollector<?, ?, ?>> collectorsList = collectors.build();
+            ImmutableList<Validator> validatorList = validators.build();
```
Comment (nmusienko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java]: copy-paste

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java [hunk 7]
```diff
@@ -0,0 +1,407 @@
...
+                this.samplesLeft = samples;
+            }
+
+            @Override
+            protected boolean terminationRequired() {
+                int s;
+                while (true) {
+                    s =  samplesLeft.get();
+                    if (s > 0) {
+                        if (samplesLeft.compareAndSet(s, --s) ) {
```
Comment (nmusienko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java]: WAT?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/PerThreadWorkloadProcess.java [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+        return totalSamplesCountRequested != -1;
+    }
+
+    @Override
+    public void changeConfiguration(WorkloadConfiguration configuration) {
+        log.debug("Configuration change request received");
+
+        for (Iterator<WorkloadService> it = threads.iterator(); it.hasNext(); ){
+            WorkloadService workloadService = it.next();
+            if (workloadService.state().equals(Service.State.TERMINATED)) {
```
Comment (nmusienko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/PerThreadWorkloadProcess.java]: what about failed threads?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/WorkloadWorker.java [hunk 9]
```diff
@@ -87,13 +90,29 @@ public String doExecute(StartWorkloadProcess command, NodeContext nodeContext) {
...
+                            (ThreadPoolExecutor) Executors.newFixedThreadPool(poolSize,
+                                    new ThreadFactoryBuilder()
+                                            .setNameFormat("workload-thread %d")
+                                            .setUncaughtExceptionHandler(ExceptionLogger.INSTANCE)
+                                            .build()
+                            ), timeoutsConfiguration);
+                } else {
+                    log.info("start per thread load process");
+                    process = new PerThreadWorkloadProcess(command.getSessionId(), command, nodeContext,
+                            (ThreadPoolExecutor) Executors.newFixedThreadPool(poolSize,
```
Comment (nmusienko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/WorkloadWorker.java]: copy-paste...

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/WorkloadWorker.java [hunk 10]
```diff
@@ -87,13 +90,29 @@ public String doExecute(StartWorkloadProcess command, NodeContext nodeContext) {
...
-                WorkloadProcess process = new WorkloadProcess(command.getSessionId(), command, nodeContext,
-                        Executors.newFixedThreadPool(poolSize,
-                                new ThreadFactoryBuilder()
-                                        .setNameFormat("workload-thread %d")
-                                        .setUncaughtExceptionHandler(ExceptionLogger.INSTANCE)
-                                        .build()
-                        ), timeoutsConfiguration);
+                WorkloadProcess process;
+
+                log.info("decide what kind of workload process should be started \nperiod = {}", command.getScenarioContext().getWorkloadConfiguration().getPeriod());
```
Comment (nmusienko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/WorkloadWorker.java]: where is decision? it is the branch

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/PerThreadWorkloadProcess.java (hunk 8 above)
Comment (amikryukov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/PerThreadWorkloadProcess.java]: I have not changed this piece of code. It have moved here from WorkloadProcess.

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java (hunk 6 above)
Comment (amikryukov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java]: Yes, but i did not refactor all logic of WorkloadServiceCreation. It moved here from WorkloadService class.

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java (hunk 7 above)
Comment (amikryukov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java]: Old code base - see WorkloadService for more details.

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/scenario/QpsClock.java [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+    }
+
+    @Override
+    public int getValue() {
+        return desiredQps.getDesiredTps().intValue();
+    }
+
+    @Override
+    public String toString() {
+        return getValue() + " rps";
```
Comment (nmusienko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/scenario/QpsClock.java]: qps?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/scenario/QpsClockConfiguration.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+        return new QpsClock(tickInterval, maxThreadNumber, createDesiredTps(BigDecimal.valueOf(getTps())));
+    }
+
+    @Override
+    public String toString() {
+        if (isRumpUp()){
+            return getTps() + " rump-up rps";
+        }
+        return getTps() + " rps";
```
Comment (nmusienko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/scenario/QpsClockConfiguration.java]: qps?

**Comment by nmusienko** [comment]
any tests? schema changes?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/PerThreadWorkloadProcess.java (hunk 8 above)
Comment (nmusienko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/PerThreadWorkloadProcess.java]: so... it's fine?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/scenario/QpsClockConfiguration.java (hunk 12 above)
Comment (amikryukov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/scenario/QpsClockConfiguration.java]: "rps" to see the same load explanation as for old implementation on WebUI and pdf report.

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java (hunk 7 above)
Comment (nmusienko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java]: Doesn't matter how old this code... Is it correct for you?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java (hunk 6 above)
Comment (nmusienko) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java]: sooo?)

**Comment by amikryukov** [comment]
There is one new test in smokes, and there are more tests in separate repo : https://github.com/amikryukov/qps-test-temporary

**Comment by nmusienko** [comment]
is new test uses periodic workload?

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/scenario/QpsClock.java (hunk 11 above)
Comment (amikryukov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/scenario/QpsClock.java]: It should be "rps" as API (if we can say so) of qps load stays the same as rps. - Same configuration, should be the same name in reports. Only implementation changed.

**Comment by amikryukov** [comment]
Yes - actually now, all tests that uses "rps-load" - uses periodic workload, as from now on "rps-load" is parsed as new clock configuration, witch send "period" in configuration to Kernels(Workers) while ticking. And that is the way workload branch is choosing:

 if (command.getScenarioContext().getWorkloadConfiguration().getPeriod() > 0) {

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/PerThreadWorkloadProcess.java (hunk 8 above)
Comment (amikryukov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/PerThreadWorkloadProcess.java]: No, but this is not the scope of my changes.

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java (hunk 6 above)
Comment (amikryukov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java]: So, I do believe that it should be done in scope of separate pull request.

File: chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java (hunk 7 above)
Comment (amikryukov) [inline chassis/core/src/main/java/com/griddynamics/jagger/engine/e1/process/AbstractWorkloadService.java]: No, but it should be done in separate pull request.
